### PR TITLE
feat: Angular 4.4 support

### LINKED ISSUE #966: Feature: Implement HttpClient
I was just wondering if there will be any implementations for the new [HttpClient](https://angular.io/guide/http) on Nativescript.

### LINKED ISSUE #1001: Angular 4.4 support
Support for angular `4.4.x`.
Related to Implement HttpClient (#966)

### PR DESCRIPTION
**Pull request by vakrilov**
In this PR:
- Bumped dependencies to `angular@4.4.x`
- Implemented `NsHttpBackEnd` that handles local requests made with `HttpClient` service
- Minor `4.4` updates

Resolves #966 
Resolves #1001

### CODE REVIEW

**Comment by vakrilov** [comment]
uitests

File: e2e/renderer/package.json [hunk 1]
```diff
@@ -6,18 +6,18 @@
   "nativescript": {
     "id": "org.nativescript.renderer",
     "tns-android": {
-      "version": "next"
+      "version": "3.2.0-2017-9-12-1"
```
Comment (sis0k0) [inline e2e/renderer/package.json]: was this intentional?

Comment (vakrilov) [reply e2e/renderer/package.json]: nope

File: nativescript-angular/http/ns-http.ts [hunk 2]
```diff
@@ -14,78 +14,78 @@ import "rxjs/add/observable/fromPromise";
...
+        }));
+    }
 }
 
 function isLocalRequest(url: string): boolean {
-  return url.indexOf("~") === 0 || url.indexOf("/") === 0;
+    return url.indexOf("~") === 0 || url.indexOf("/") === 0;
 }
 
 function normalizeLocalUrl(url: string): string {
```
Comment (sis0k0) [inline nativescript-angular/http/ns-http.ts]: This method can be reused in `NsHttpBackEnd`

File: nativescript-angular/http/ns-http.ts [hunk 3]
```diff
@@ -14,78 +14,78 @@ import "rxjs/add/observable/fromPromise";
...
+                    reject(responseOptions(err, 400, url));
+                });
+            } else {
+                reject(responseOptions("Not Found", 404, url));
+            }
+        }));
+    }
 }
 
 function isLocalRequest(url: string): boolean {
```
Comment (sis0k0) [inline nativescript-angular/http/ns-http.ts]: This is duplicated in `NsHttpBackEnd`. May be a good idea to extract it in a common module?

File: nativescript-angular/http-client/ns-http-backend.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        return result;
+    }
+
+    private getAbsolutePath(url: string): string {
+        url = url.replace("~", "").replace("/", "");
+        url = path.join(this.nsFileSystem.currentApp().path, url);
+        return url;
+    }
+
+    private handleLocalRequest(url: string): Observable<HttpEvent<any>> {
```
Comment (sis0k0) [inline nativescript-angular/http-client/ns-http-backend.ts]: This seems to be the same as `_requestLocalUrl` from `NSHttp` except the additional error handling for failed parsing. Can we move the method to a common module (with the error handling)?

File: nativescript-angular/http-client/ns-http-backend.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        return new Observable((observer: Observer<HttpEvent<any>>) => {
+            if (this.nsFileSystem.fileExists(url)) {
+                const localFile = this.nsFileSystem.fileFromPath(url);
+                localFile.readText()
+                    .then((data) => {
+                        try {
+                            const json = JSON.parse(data);
+                            observer.next(createSuccessResponse(url, json, 200));
+                            observer.complete();
+                        } catch (error) {
```
Comment (sis0k0) [inline nativescript-angular/http-client/ns-http-backend.ts]: Catch only `SyntaxError`s?

File: nativescript-angular/http-client/http-client.module.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// IMPORTant: Importing "@angular/common/http" for the first time overwrites the
+// global.__extends function.
+const cachedExtends = global.__extends;
+import { HttpClientModule, HttpBackend } from "@angular/common/http";
+global.__extends = cachedExtends;
+
+import { NSFileSystem } from "../file-system/ns-file-system";
+import { NsHttpBackEnd } from "./ns-http-backend";
+
+export { NsHttpBackEnd } from "./ns-http-backend";
```
Comment (sis0k0) [inline nativescript-angular/http-client/http-client.module.ts]: Why is that reexported from here?

File: ng-sample/app/examples/http-client/http-client-test.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
+import { Component, Inject, Injectable } from "@angular/core";
+import {
+    HttpClient, HTTP_INTERCEPTORS, HttpEventType, HttpErrorResponse,
+    HttpEvent, HttpInterceptor, HttpHandler, HttpRequest
+} from "@angular/common/http";
+import { Observable } from "rxjs/observable";
```
Comment (sis0k0) [inline ng-sample/app/examples/http-client/http-client-test.ts]: `from "rxjs/observable"` -> `from "rxjs/Observable"`

Comment (vakrilov) [reply ng-sample/app/examples/http-client/http-client-test.ts]: Nice catch

**Comment by SvetoslavTsenov** [comment]
e2e

**Comment by vakrilov** [comment]
e2e

File: nativescript-angular/http-client/ns-http-backend.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        if (isLocalRequest(req.url)) {
+            result = this.handleLocalRequest(req.url);
+        } else {
+            result = super.handle(req);
+        }
+
+        return result;
+    }
+
+    private handleLocalRequest(url: string): Observable<HttpEvent<any>> {
```
Comment (sis0k0) [inline nativescript-angular/http-client/ns-http-backend.ts]: The name of the method is a bit confusing as it matches the name of the imported function.

File: nativescript-angular/http/ns-http.ts [hunk 9]
```diff
@@ -11,81 +11,62 @@ import {
...
+        if (isLocalRequest(url)) {
+            return this.handleLocalRequest(url);
+        } else {
+            return super.get(url, options);
+        }
+    }
 
-function normalizeLocalUrl(url: string): string {
-  return url.replace("~", "").replace("/", "");
+    private handleLocalRequest(url: string): Observable<Response> {
```
Comment (sis0k0) [inline nativescript-angular/http/ns-http.ts]: Same as above

**Comment by vakrilov** [comment]
e2e

**Comment by vchimev** [comment]
renderer-android

**Comment by jogboms** [comment]
@vakrilov 

ERROR Error: Uncaught (in promise): Error: com.tns.NativeScriptException: Failed to find module: "nativescript-angular/value-accessors/base-value-accessor", relative to: app/tns_modules/

Just installed `@next`

`@angular/*: "^4.4.1"`

**Comment by jogboms** [comment]
I can see `nativescript-angular/value-accessors` has been moved to `nativescript-angular/forms/value-accessors`

**Comment by jogboms** [comment]
`nativescript-angular/common/utils` no longer exists as well

**Comment by vakrilov** [comment]
Hey @jogboms, thanks for reporting that!
Actually, the `nativescript-angular/common/utils` was removed a few releases back (before `4.x`).
We heave moved the `value-accessors` inside the forms in this PR(#1003) which is a breaking change for plugins/apps that import it. We are currently working on a patch with a fix.

**Comment by jogboms** [comment]
Okay. Didn't notice that. Just sent a PR over.

**Comment by vchimev** [comment]
@jogboms,
`nativescript-angular@4.4` does not require `tns-core-modules@next`, however it works with both *latest* and *next* version.

**Comment by jogboms** [comment]
Very well @vchimev. Thanks.

**Comment by vchimev** [comment]
`nativescript-angular@4.4.1` containing a compatibility fix has been released - here is the [changelog](https://github.com/NativeScript/nativescript-angular/blob/master/CHANGELOG.md#441-2017-10-13).

**Comment by jogboms** [comment]
Nice one. 👍 @vchimev

**Pull request merged**